edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/KubeUtils.cs

```diff
@@ -0,0 +1,219 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Agent.Kubernetes
+{
+    using System;
+    using System.Linq;
+    using System.Text;
+    using CoreConstants = Microsoft.Azure.Devices.Edge.Agent.Core.Constants;
+
+    public static class KubeUtils
+    {
+        const string Alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";
+        const string AllowedCharsLabelValues = "-._";
+        const string AllowedCharsDns = "-";
+        const string AllowedCharsGeneric = "-.";
+        const int MaxK8SValueLength = 253;
+        const int MaxDnsNameLength = 63;
+        const int MaxLabelValueLength = 63;
+
+        static bool IsAlpha(char ch) => Alphabet.IndexOf(ch) != -1;
+
+        static bool IsAlphaNumeric(char ch) => IsAlpha(ch) || Char.IsDigit(ch);
+
+
+        // Valid annotation keys have two segments: an optional prefix and name, separated by a slash (/). 
+        // The name segment is required and must be 63 characters or less, beginning and ending with an 
+        // alphanumeric character ([a-z0-9A-Z]) with dashes (-), underscores (_), dots (.), and alphanumerics between. 
+        // The prefix is optional. If specified, the prefix must be a DNS subdomain
+        public static string SanitizeAnnotationKey(string key)
+        {
+            if (string.IsNullOrEmpty(key))
+            {
+                throw new InvalidKubernetesNameException($"Key '{key}' is null or empty");
+            }
+
+            char[] annotationSplit = { '/' };
+            string[] keySegments = key.Split(annotationSplit, 2);
+            if (keySegments.Count() == 2)
+            {
+                return SanitizeDNSDomain(keySegments[0]) + "/" + SanitizeNameValue(keySegments[1]);
+            }
+            else
+            {
+                return SanitizeNameValue(key);
+            }
+        }
+
+        // Alphanumeric, '-' and '.' up to 253 characters.
+        public static string SanitizeK8sValue(string name)
+        {
+            if (name == null)
+            {
+                // Values sometimes may be null, and that's OK.
+                return name;
+            }
+
+            name = name.ToLower();
+
+            var output = new StringBuilder();
+            for (int i = 0; i < name.Length; i++)
+            {
+                if (IsAlphaNumeric(name[i]) || AllowedCharsGeneric.IndexOf(name[i]) != -1)
+                {
+                    output.Append(name[i]);
+                }
+            }
+
+            if (output.Length > MaxK8SValueLength)
+            {
+                throw new InvalidKubernetesNameException($"Value '{name}' as sanitized exceeded maximum length {MaxK8SValueLength}");
+            }
+
+            return output.ToString();
+        }
+
+        // DNS label (as per RFC 1035)
+        public static string SanitizeDNSValue(string name, int maxLength = MaxDnsNameLength)
```

Comment:
Do we want to allow the caller of this method to pass `maxLength` in the method? Now, this method is somewhat generic and not very specific to k8s, right?

Reply:
This is an artifact from a previous iteration.  I should remove this.